File: server/django/record/sgf.py
from sgfmill import sgf

from record.models import Record


def export_sgf(record: Record) -> bytes:
    game = sgf.Sgf_game(size=record.board_size)
    root_node = game.get_root()
    if record.name:
        root_node.set('GN', record.name)
    if record.comment:
        root_node.set('GC', record.name)
    root_node.set('PB', record.black_player)
    root_node.set('PW', record.white_player)
    root_node.set('KM', record.komi)
    root_node.set('HA', record.handicap)
    if record.winner != Record.Winner.UNDECIDED:
        root_node.set('RE', f'{record.winner}+R')
    else:
        root_node.set('RE', 'Void')

    for move in record.moves.all():
        node = game.extend_main_sequence()
        # position increments left to right, then top to bottom
        # sgfmill considers (0,0) to be the lower left corner
        # Therefore, we need to invert y
        if move.position is not None:
            x = move.position % record.board_size
            y = record.board_size - (move.position // record.board_size) - 1
            # and also flip x/y??? whatever dude
            node.set_move(move.color.lower(), (y, x))
        else:
            node.set_move(move.color.lower(), None)
    return game.serialise()

File: server/haskell/src/SGF.hs
module SGF (Game (..), toSGF) where

import Record.Go (Color (Black, White), Move, Pos, toCoord')

data Game = Game
  { name :: Maybe String,
    comment :: Maybe String,
    handicap :: Int,
    komi :: Double,
    blackPlayer :: String,
    whitePlayer :: String,
    winner :: Maybe Color,
    size :: Int,
    moves :: [Move]
  }

alphabet :: String
alphabet = "abcdefghijklmnopqrst"

formatPos :: Int -> Maybe Pos -> String
formatPos _ Nothing = "tt"
formatPos size (Just pos) = (alphabet !! x) : [alphabet !! y]
  where
    (x, y) = toCoord' size pos

formatMove :: Int -> Move -> String
formatMove size (pos, Black) = "B[" ++ formatPos size pos ++ "]"
formatMove size (pos, White) = "W[" ++ formatPos size pos ++ "]"

toSGF :: Game -> String
toSGF Game {name, comment, handicap, komi, blackPlayer, whitePlayer, winner, size, moves} =
  "(;FF[4]CA[UTF-8]"
    ++ ( case comment of
           Just comment' -> "GC[" ++ comment' ++ "]"
           Nothing -> ""
       )
    ++ "GM[1]"
    ++ ( case name of
           Just name' -> "GN[" ++ name' ++ "]"
           Nothing -> ""
       )
    ++ ("HA[" ++ show handicap ++ "]")
    ++ ("KM[" ++ show komi ++ "]")
    ++ ("PB[" ++ blackPlayer ++ "]")
    ++ ("PW[" ++ whitePlayer ++ "]")
    ++ ( "RE["
           ++ ( case winner of
                  Nothing -> "Void"
                  Just Black -> "B+R"
                  Just White -> "W+R"
              )
           ++ "]"
       )
    ++ ("SZ[" ++ show size ++ "]")
    ++ foldl
      (\a b -> a ++ ";" ++ b)
      ""
      (map (formatMove size) moves)
    ++ ")"
